### PR TITLE
Single-task POYO

### PR DESCRIPTION
**Pull request by vinamarora8**
Tested on entire MP dataset for up to 60 epochs. Works as expected. On 4xA40s: time per epoch: ~40s, time per validation: ~10s. I believe we are ready for training POYO-MP.

There is some ugliness wrt `subtask_index` handling. But I am deliberately leaving that until brainsets merges the interval-based approach so we don't have to re-engineer it later.

### CODE REVIEW

File: examples/poyo/configs/base.yaml
Comment (vinamarora8) [inline examples/poyo/configs/base.yaml:29]: This is hacky. Will clean once the interval-based movement phases are merged into brainsets.

**Comment by vinamarora8** [comment]
@mazabou Ready for review. If possible, we should merge this in before adding `examples/poyo/finetune.py`

**Review by mazabou — COMMENTED** [review]
Looks good overall. I left a few comments. 

I made the update to `brainsets`, I can attempt to incorporate them here.

File: examples/poyo/README.md
Comment (vinamarora8) [reply examples/poyo/README.md:10]: Yes, we should remember to do that

File: examples/poyo/configs/base.yaml
Comment (vinamarora8) [reply examples/poyo/configs/base.yaml:11]: oh, I just copied these from poyo_plus I think. Although I added them there. Not needed for supervised cases. I'll remove it

Comment (mazabou) [reply examples/poyo/configs/base.yaml:11]: yes, it's not you who added it. I meant to ask this for a while. I know we had it before.

File: torch_brain/utils/stitcher.py
Comment (vinamarora8) [reply torch_brain/utils/stitcher.py:77]: It is specific to multi-session situations because it requires the tokenizer to add `"session_id"` field to the batch-data. In single session situations users shouldn't be expected to add this field.

I believe having the `Stitcher` class (independent of Lightning, the idea we discussed on slack) would end up making this situation more unified. This would take some thinking and experimenting; which I can probably get to next week, but this week I'm tight on time. Until then, I feel this is a good solution.

Comment (mazabou) [reply torch_brain/utils/stitcher.py:77]: I agree with the second point. 

Re: `"session_id"`, I also agree. My solution is that instead of rewritting the whole class, we just add an argument to the existing one that would be "single_session=True". And it will not expect to get "session_id" in that case. everything else will be exactly the same.

File: torch_brain/utils/stitcher.py
Comment (vinamarora8) [inline torch_brain/utils/stitcher.py:229]: Note to self: change the name to `MultiSessionMultiTaskDecodingStitchEvaluator`

File: torch_brain/utils/stitcher.py
Comment (vinamarora8) [reply torch_brain/utils/stitcher.py]: Removed "MultiSession" from the name

File: torch_brain/models/poyo.py
Comment (mazabou) [inline torch_brain/models/poyo.py:329]: this will be replaced so I will not review it

File: torch_brain/models/poyo.py
Comment (mazabou) [inline torch_brain/models/poyo.py:253]: should we name it `readout_spec`? `modality_spec` is more consistent, but could be confusing since spikes itself is a modality. and in the future we can change the input modality. so for consistency I would suggest calling this `readout_spec`

File: torch_brain/models/poyo.py
Comment (mazabou) [inline torch_brain/models/poyo.py:346]: here padding is used instead of chaining. how are you dealing with the loss when it comes to padding values?

**Review by mazabou — COMMENTED** [review]
Review of `torch_brain/models`

File: examples/poyo/train.py
Comment (mazabou) [inline examples/poyo/train.py:38]: the python convention is that `main()` goes at the bottom of the script. so this should be the last function.

File: examples/poyo/train.py
Comment (mazabou) [inline examples/poyo/train.py:190]: why are we going backwards here?

File: torch_brain/utils/callbacks.py
Comment (mazabou) [inline torch_brain/utils/callbacks.py:23]: why do we need to sync_dist here? isn't everything blocked until the last gpu is done? also, why does this need to be accurate?

File: torch_brain/utils/stitcher.py
Comment (mazabou) [inline torch_brain/utils/stitcher.py:118]: what's `metric_factory`? this nullifies the option to specify a different metric per dataset?

Comment (mazabou) [reply torch_brain/utils/stitcher.py:118]: are we creating a concept that will not work for POYOPlus?

File: examples/poyo/configs/train_poyo_mp.yaml
Comment (mazabou) [inline examples/poyo/configs/train_poyo_mp.yaml:15]: why mini?

**Review by mazabou — COMMENTED** [review]
Review part 2 of 2

File: torch_brain/models/poyo.py
Comment (vinamarora8) [reply torch_brain/models/poyo.py:253]: We should change this in the spec python file as well in that case. I just went with whatever name was used there. I support renaming it to readout_spec at both places.

File: examples/poyo/train.py
Comment (vinamarora8) [reply examples/poyo/train.py:38]: yeah we can rename this to `train()` or smth. Otherwise I don't like having both DataModule and TrainWrapper at the top. Too much code before you get to what this script is actually running. Happy to reconsider.

File: examples/poyo/train.py
Comment (vinamarora8) [reply examples/poyo/train.py:190]: Just wanted to get it done without having to test _this_ code on datasets with other kinds of losses. We can bring back other losses when we've tested them. Imo we shouldn't include a feature we haven't tested. Adding more losses can be separate PRs.

File: torch_brain/utils/callbacks.py
Comment (vinamarora8) [reply torch_brain/utils/callbacks.py:23]: It's not needed in theory. Lightning complains if we don't have sync_dist=True. I'd rather not see that warning. This incurs negligible cost.

File: torch_brain/utils/stitcher.py
Comment (vinamarora8) [reply torch_brain/utils/stitcher.py:118]: `metric_factory` is when you wanna use a custom metric bypassing the modality_spec framework.

Comment (vinamarora8) [reply torch_brain/utils/stitcher.py:118]: This will not work for POYOPlus. This stitcher is not meant for handling multi-task output

Comment (vinamarora8) [reply torch_brain/utils/stitcher.py:118]: I think I have made a mistake here. metric_factor's type should be `Optional[Callable[[int], torchmetrics.Metric]`

File: examples/poyo/configs/train_poyo_mp.yaml
Comment (vinamarora8) [reply examples/poyo/configs/train_poyo_mp.yaml:15]: Idk, ill call it poyo_mp

File: torch_brain/models/poyo.py
Comment (vinamarora8) [reply torch_brain/models/poyo.py:39]: Sure, it can be a separate PR focused on POYOPlus edits.

File: torch_brain/models/poyo.py
Comment (vinamarora8) [reply torch_brain/models/poyo.py]: done

File: torch_brain/models/__init__.py
Comment (vinamarora8) [reply torch_brain/models/__init__.py:1]: done

**Comment by vinamarora8** [comment]
Note I have removed `strategy=ddp_find_unused_parameters_true ...` and `accelerator = ...` arguments of lightning trainer because we don't need them for this code.

File: torch_brain/models/poyo.py
Comment (mazabou) [reply torch_brain/models/poyo.py]: To clarify, I am not referring to chain vs. pad. What I am suggesting is replicating the `unpack_output=True` behavior, so that instead of returning `B x T_out x D_out`, we return `List[*, D_out]` of length `B`. The logic for this is super simple. The typical user is not aware that masking is being used! — it's happening in the tokenizer/collate. so they wouldn't know that the output they are getting will need to be "unpadded", in which case they would have to know how to access to the `output_mask` inside of the batch dict. This would of course happen rarely but it is likely incase some dataset has missing values etc...

My suggestion is to include a single line of code:
```
if unpack_output:
     return [output[b][output_mask[b]] for b in range(output.size(0))]
```

File: torch_brain/models/poyo.py
Comment (mazabou) [reply torch_brain/models/poyo.py:253]: what do you mean by the spec python file? you mean renaming `ModalitySpec` to `ReadoutSpec`?

File: examples/poyo/train.py
Comment (mazabou) [reply examples/poyo/train.py:38]: let's put it back the way it was please. if you want to make the change across the board, it should be a separate PR/ a separate discussion.

File: examples/poyo/train.py
Comment (mazabou) [reply examples/poyo/train.py:186]: we have `compute_loss_or_metric` that we can use no?

File: torch_brain/utils/stitcher.py
Comment (mazabou) [reply torch_brain/utils/stitcher.py]: this is a new concept. I would rather it be in a separate PR/discussion. For now, I do not see why are introducing two completely different ways for using the stitcher, based on whether it's POYO or POYOPlus. It will be extremely confusing.

Comment (vinamarora8) [reply torch_brain/utils/stitcher.py]: is your issue with the existence of `metric_factory` as an optional parameter?

File: examples/poyo/train.py
Comment (vinamarora8) [reply examples/poyo/train.py]: forgot about compute_loss_or_metric. fixed.

File: torch_brain/models/poyo.py
Comment (vinamarora8) [reply torch_brain/models/poyo.py:253]: Yes

File: torch_brain/models/poyo.py
Comment (vinamarora8) [reply torch_brain/models/poyo.py]: Fair point. Done.

File: torch_brain/utils/stitcher.py
Comment (vinamarora8) [reply torch_brain/utils/stitcher.py]: removed

File: torch_brain/models/poyo.py
Comment (vinamarora8) [reply torch_brain/models/poyo.py]: Renamed `modality_spec` to `readout_spec`. Done.

**Review by mazabou — APPROVED** [review]
Look great! Thank you for the colossal amount of work. Good to merge.

Just checking, you are going to merge this into main correct?

**Pull request merged**
